modules/system/assets/js/snowboard/ajax/Request.js

```diff
@@ -0,0 +1,789 @@
+                    elements.forEach((element) => {
+                        const sanitizedContent = this.snowboard.sanitizer().sanitize(content);
+
+                        switch (mode) {
+                            case 'append':
+                                element.innerHTML += sanitizedContent;
+                                break;
+                            case 'prepend':
+                                element.innerHTML = sanitizedContent + element.innerHTML;
+                                break;
+                            case 'replace':
+                            default:
+                                element.innerHTML = sanitizedContent;
+                                break;
+                        }
+
+                        affected.push(element);
+
+                        // Fire update event for each element that is updated
+                        this.snowboard.globalEvent('ajaxUpdate', element, sanitizedContent, this);
+                        const event = new Event('ajaxUpdate');
+                        event.content = sanitizedContent;
+                        element.dispatchEvent(event);
+                    });
+                }
+            });
+
+            this.snowboard.globalEvent('ajaxUpdateComplete', affected, this);
+
+            resolve();
+        });
+    }
+
+    /**
+     * Processes the response data.
+     *
+     * This fires off all necessary processing functions depending on the response, ie. if there's any flash
+     * messages to handle, or any redirects to be undertaken.
+     *
+     * @param {Object} response
+     * @returns {void}
+     */
+    processResponse(response) {
+        if (this.options.success && typeof this.options.success === 'function') {
+            if (!this.options.success(this.responseData, this)) {
+                return;
+            }
+        }
+
+        // Allow plugins to cancel any further response handling
+        if (this.snowboard.globalEvent('ajaxSuccess', this.responseData, this) === false) {
+            return;
+        }
+
+        // Allow the element to cancel any further response handling
+        if (this.element) {
+            const event = new Event('ajaxDone', { cancelable: true });
+            event.responseData = this.responseData;
+            event.request = this;
+            this.element.dispatchEvent(event);
+
+            if (event.defaultPrevented) {
+                return;
+            }
+        }
+
+        // Check for a redirect from the response, or use the redirect as specified in the options. This takes
+        // precedent over all other checks.
+        if (this.redirect || response.X_WINTER_REDIRECT) {
+            this.processRedirect(this.redirect || response.X_WINTER_REDIRECT);
+            return;
+        }
+
+        if (this.flash && response.X_WINTER_FLASH_MESSAGES) {
+            this.processFlashMessages(response.X_WINTER_FLASH_MESSAGES);
+        }
```

Comment:
Should we process flash messages before processing redirects in case they display a flash message before the redirect? Also would help with redirects to file downloads. 

I ask because if the flash message has been sent then it's already been pulled from the session and theoretically won't be sent again. Most useful feature would be something like returning a redirect to a file download and displaying a flash message "Downloading file..."

Reply:
Yeah, makes sense.